logstash-core/lib/logstash/instrument/metric_store.rb

```diff
@@ -28,10 +32,20 @@ def initialize
     # path in the hash and return `new_value`
     #
     # @param [Array] The path where the values should be located
-    # @param [Object] The default object if the value is not found in the path
+    # @param [Symbol] The metric key
     # @return [Object] Return the new_value of the retrieve object in the tree
     def fetch_or_store(namespaces, key, default_value = nil)
-      fetch_or_store_namespaces(namespaces).fetch_or_store(key, block_given? ? yield(key) : default_value)
+      provided_value =  block_given? ? yield(key) : default_value
+
+      # Check if the value is in the `fast_lookup` if it is return it, if not compute it
+      # and insert it into the store for
+      if found_value = @fast_lookup.put_if_absent([namespaces, key], provided_value)
```

Comment:
```
We first check in the `@fast_lookup` store to see if we have already see that metrics before, This give us a `o(1)` access, which is faster than searching through the structured data store (Which is a `o(n)` operation where `n` is the number of element in the namespace and the value of the key). If the metric is already present in the `@fast_lookup`, the call  to `#put_if_absent` will return the value. This value is send back directly to the caller.

BUT. If the value is not present in the `#fastlookup` the value will be inserted and `#puf_if_absent` will return nil.
With this returned value  of `nil` we assume that we don't have it in the `@metric_store` for structured search so we add it there too.

The problem with only using the `@metric_store` directly all the time would require us to use the mutex around the structure since its a multi-level hash, without that it wont return a consistent value of the metric and this would slow down the code.
```

@purbon  can you review this comment ^?